devito/passes/iet/orchestration.py

```diff
@@ -177,45 +60,43 @@ def _make_waitlock(self, iet, sync_ops, *args):
     def _make_withlock(self, iet, sync_ops, pieces, root):
         locks = sorted({s.lock for s in sync_ops}, key=lambda i: i.name)
 
-        threads = self.__make_threads(value=min(i.size for i in locks))
+        npthreads = min(i.size for i in locks)
```

Comment:
Can you walk me through this piece if code as it's not clear to my why this gives the number of npthreads. And why is locks sorted?

Reply:
Ah, `locks` is sorted for deterministic code gen. I shall add a comment

About the `npthreads` -- in practice, this routine constructs the piece of tree responsible for the copy-back from the device to the host. For a copy-back to be feasile all of the locks protecting the copied fileds must have been acquired. The `min` is  used to pick the maximum degree of parallelism possible. For example, consider:

```
while (lock0[i] == 0 && lock1[j] == 0);  // busy wait
```

where `i` and `j` are modulo dimensions, `lock0` protects 3 entries (say `u[0], u[1], u[2]`), and `lock1` protects 2 entries (say `v[0], v[1]`). There can't be more than two `v`'s in flight at the same time, so it's pointless to assume that more than 2 threads will ever be needed. I will also add a comment